### PR TITLE
8287812: Cleanup JDWP agent GetEnv initialization

### PR DESCRIPTION
**Pull request by plummercj**
3 things to cleanup in this area:

(1) The JDWP agent uses `JNI GetEnv(JVMTI_VERSION_1)` to get a JVMTI environment. If `GetEnv` fails the JDWP agent prints this:

`ERROR: JDWP unable to access JVMTI Version 1 (0x30010000), is your J2SE a 1.5 or newer version? JNIEnv's GetEnv() returned -3`

The text "is your J2SE a 1.5 or newer version?" dates from JDK 5 when JVMTI was introduced and doesn't make sense now.

(2) `JVMTI_VERSION_1` suggests that the JDWP agent is looking for a JVMTI v1 environment when it really wants the latest. `GetEnv` should request `JVMTI_VERSION` so that it always requests the current version.

(3) There is some outdated compatibility checking between runtime and compile time versions of JVMTI that date back to the 1.0, 1.1, and 1.2 era, and are no longer needed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8287812](https://bugs.openjdk.org/browse/JDK-8287812): Cleanup JDWP agent GetEnv initialization

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11602/head:pull/11602` \
`$ git checkout pull/11602`

Update a local copy of the PR: \
`$ git checkout pull/11602` \
`$ git pull https://git.openjdk.org/jdk pull/11602/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11602`

View PR using the GUI difftool: \
`$ git pr show -t 11602`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11602.diff">https://git.openjdk.org/jdk/pull/11602.diff</a>

</details>

### CODE REVIEW

File: src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c [hunk 1]
```diff
@@ -172,23 +172,6 @@ static jboolean
 compatible_versions(jint major_runtime,     jint minor_runtime,
                     jint major_compiletime, jint minor_compiletime)
 {
-    /*
-     * First check to see if versions are explicitly compatible via the
-     * list specified above.
-     */
-    int i;
-    for (i = 0; i < nof_compatible_versions; ++i) {
```
Comment (alexmenkov) [inline src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:180]: nof_compatible_versions and compatible_versions_list variables are not used now and can be deleted

**Comment by dholmes-ora** [comment]
> (2) JVMTI_VERSION_1 suggests that the JDWP agent is looking for a JVMTI v1 environment when it really wants the latest. 

Does it? I guess it is a somewhat moot point inside the same VM but don't you only need to ask for the version of the environment that supports the API you are going to use?

File: src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c (hunk 1 above)
Comment (plummercj) [reply src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:180]: Yes, I meant to delete all that. Fixed now.

**Comment by plummercj** [comment]
> > (2) JVMTI_VERSION_1 suggests that the JDWP agent is looking for a JVMTI v1 environment when it really wants the latest.
> 
> Does it? I guess it is a somewhat moot point inside the same VM but don't you only need to ask for the version of the environment that supports the API you are going to use?

Yes, but why would we need to allow the debug agent to bind to a version of JVMTI other than the one it was built with? I think there might be some historical context here that I'm missing. I know in the past there has been some conditional code in the debug agent based on what features JVMTI supports. For example, SuspendThreadList. I removed the code that was needed if SuspendThreadList was not supported, and we now assume that it is supported. Maybe the debug agent built by Oracle (probably actually Sun at the time) was also used by other JVMs, and was expected to work with older versions of JVMTI. I doubt that is the case anymore.

**Comment by dholmes-ora** [comment]
> Maybe the debug agent built by Oracle (probably actually Sun at the time) was also used by other JVMs, and was expected to work with older versions of JVMTI. I doubt that is the case anymore.

I don't know what the deployment model for the debug agent is, but you have now made it so that it can't work with older versions. Does this affect the ability to run an earlier version of the debugger VM compared to the debugee VM?

**Comment by plummercj** [comment]
> I don't know what the deployment model for the debug agent is, but you have now made it so that it can't work with older versions. Does this affect the ability to run an earlier version of the debugger VM compared to the debugee VM?

I don't believe so. The debugger can query the debug agent's capabilities (and most do so to support older versions). The debug agent imposes no type of versioning or capabilities on the debugger:

https://docs.oracle.com/en/java/javase/17/docs/specs/jdwp/jdwp-protocol.html#JDWP_VirtualMachine_CapabilitiesNew

**Comment by sspitsyn** [comment]
Should we consider filing a CSR as potentially there is a minimal compatibility risk.

**Comment by AlanBateman** [comment]
In the distant past it may have been useful to copy libjdwp/libsocket_dt from a JDK build into a build (JRE or embedded) that didn't have the debugger agent. Those days are long gone. Also since JDK 9, the JDK build generates a hash for each packaged module that is later use by jlink to detect any accidental mix 'n matching of modules from one JDK build with another. This will catch accidental attempts to use the jdk.jdwp.agent module one from build with java.base from another build.

The error message "is your J2SE a 1.5 or newer version?" has come several times during JDK 19 and JDK 20 development, mostly in GHA with new tests attempting to run with --enable-preview and use JVMTI (jshell tests for example). JDK-8287812 was created to remove the confusing part of the error message but has since expanded to broader cleanup, which is okay.  If virtual threads become permanent in JDK 21, and if the are outstanding ports of VM continuations, then it means that JVMTI may have to be disabled completely for those ports (in which case the error in PR will be seem widely, unless the tests are changed to skip) or JVMTI will have to be changed to work the alternative implementation of virtual threads (in which case this error should not be seen).

**Review by AlanBateman — APPROVED** [review]

**Comment by plummercj** [comment]
> Should we consider filing a CSR as potentially there is a minimal compatibility risk?

It don't think it should be necessary. I don't believe there are any known compatibility issues here. Although we switched from requesting `JVMTI_VERSION_1` to requesting the current compile time `JVMTI_VERSION`, it's always been a requirement of the `compatible_versions()` check that the compile time and runtime major versions match up. What has changed somewhat is that the micro versions must also match up, but I don't think we bother with minor or micro versions anymore.

There is one downside of requesting `JVMTI_VERSION` instead of `JVMTI_VERSION_1` that I'm now realizing. If you do try to drop a newer debug agent into an older jdk release, GetEnv() will fail. That's ok, but if we allowed it to succeed by fetching `JVMTI_VERSION_1`, then instead the `compatible_versions()` check would fail, which would give you a more meaningful error message. In other words, rather than a "GetEnv() failed" message you would get a "incompatible version" error message.

**Comment by dcubed-ojdk** [comment]
I have a memory of the following concerning JVM/TI versions:

- if you asked for JVMTI_VERSION, then you got the highest version supported by the JVM.
- if you asked for JVMTI_VERSION_1, then you got the highest compatible version to VERSION_1.
- if you asked for JVMTI_VERSION_1_1, then you got JVMTI_VERSION_1_1, no more, no less if the JVM supports it.

Here's where things get gnarly:
- if you asked for JVMTI_VERSION_1, could you get back JVMTI_VERSION_1_2 or JVMTI_VERSION_9 or
  JVMTI_VERSION_11 if the JVM in question supported each of those versions (and no higher)?
  - the assumption that I've always made is that each of JVMTI_VERSION_1_N are all considered
    compatible so asked for JVMTI_VERSION_1, you would get the highest compatible version in
    the JVMTI_VERSION_1_N set that is supported by the JVM on which you ask the question.
  - if you ask for JVMTI_VERSION_1 on a JVM that supports JVMTI_VERSION_1_N and JVMTI_VERSION_9
    then I would expect you to get back JVMTI_VERSION_1_2 and not JVMTI_VERSION_9. Why?
  - My assumption is that you only change major version numbers when you make incompatible
    changes. Adding a new API is a compatible change because the older agent doesn't know how
    to use the new API. Changing the semantics of an existing API is NOT a compatible change so
    you bump the major number.

So I'm assuming that JVMTI_VERSION_9 has incompatible changes relative to JVMTI_VERSION_1_N.
Similarly I'm assuming that JVMTI_VERSION_11 has incompatible changes relative to JVMTI_VERSION_9
and JVMTI_VERSION_1_N.

**Comment by dcubed-ojdk** [comment]
Of course, it is possible that JVMTI_VERSION_9 is compatible with JVMTI_VERSION_1_N
and JVMTI_VERSION_11 is compatible with both JVMTI_VERSION_9 and JVMTI_VERSION_1_N.
We might have simply bumped the major number because we were in that train of thought
back in the JDK9 days...

I have not checked the compatibility matrices for the various JVM/TI releases in a very, very
long time.

**Comment by plummercj** [comment]
> if you asked for JVMTI_VERSION_1, could you get back JVMTI_VERSION_1_2 or JVMTI_VERSION_9 or
JVMTI_VERSION_11 if the JVM in question supported each of those versions (and no higher)? 

Yes, but the debug agent does a version check to make sure the runtime version matches the compile time version.

Right now if I ask for JVMTI_VERSION_1 I will get back version 21, so it seems that GetEnv() always returns the version that was current when JVMTI was built, except of course it will fail if the requested version is newer.

It seems we update the major version with every JDK release. Here's what currently is produced in jvmti.h:

enum {
    JVMTI_VERSION_1   = 0x30010000,
    JVMTI_VERSION_1_0 = 0x30010000,
    JVMTI_VERSION_1_1 = 0x30010100,
    JVMTI_VERSION_1_2 = 0x30010200,
    JVMTI_VERSION_9   = 0x30090000,
    JVMTI_VERSION_11  = 0x300B0000,
    JVMTI_VERSION_19  = 0x30130000,
    JVMTI_VERSION = 0x30000000 + (21 * 0x10000) + ( 0 * 0x100) + 0  /* version: 21.0.0 */
};

So it seems right now there is no difference in the end result between using JVMTI_VERSION and JVMTI_VERSION_<n>, assuming the compile and runtime versions match (you aren't dropping the debug agent into a different JDK build).

Now for jvmti agents that possibly could end up running in various JDK releases, the difference is more meaningful. They probably do not want to use JDK_VERSION because this could cause the agent to fail to initialize with older JDK releases that the agent author intends for agent to work with. Instead they should specify the minimum JVMTI version that they require.

Having said that it may at first seem odd that the debug agent would want to request JVMTI_VERSION_1, since it actually requires something newer, but the check to make sure that the runtime and compile time versions are the same will produce an error if indeed the runtime version is older than what we built against.  So the advantage of requesting an old version and then doing a runtime version check is that is that you can produce a very specific error message if the versioning is not right. If you rely on GetEnv to catch versioning issues, all you know is that you got a JNI_EVERSION error code.

**Comment by AlanBateman** [comment]
> I have a memory of the following concerning JVM/TI versions:
> 
> * if you asked for JVMTI_VERSION, then you got the highest version supported by the JVM.

JVMTI_VERSION is defined in jvmti.h so its value depends on which JDK include directory was used when compiling the agent. If the JDK supports that version then it is required to return a JVMTI env that is compatible or fail with EVERSION. If the agent calls GetVersionNumber to see what version it got then it may be a newer (but compatible) version, which is what our implementation does. It may be that the agent was compiled with JVMTI_VERSION for a newer JDK release, in which case GetEnv must fail with EVERSION as an older JDK release doesn't know about new JVMTI versions.

In any case, there should be no mix'ing and matching with the JDWP agent so GetEnv asking for JVMTI_VERSION should be okay. For a JDK build, the JVMTI version at build time will match run-time. A hash of the jdk.jdwp.agent module is generated at build/packaging time to prevent accidental linking with modules from different JDK builds. If someone does attempt to run jlink and use java.base from one build and jdk.jdwp.agent from another build then they will get a error that the hashes don't match. It is of course possible that someone does a slight of hand and copy a libjdwp/libdt_socket from one JDK build into a build from a different JDK release but I don't think we need to spend too much time on that as it's just not supported to do that.


> * if you asked for JVMTI_VERSION_1, then you got the highest compatible version to VERSION_1.
> * if you asked for JVMTI_VERSION_1_1, then you got JVMTI_VERSION_1_1, no more, no less if the JVM supports it.

If you run on JDK 19 then you'll get a JVMTI_VERSION_19 in both cases. It might be in the future that we need to do some incompatible changes to JVMTI, like remove the deprecated heap functions, in which case it might have to return EVERSION for both cases. The is a bridge that we haven't got to yet but I expect it will require discussing JVMTI capabilities at the same time.

**Comment by plummercj** [comment]
@AlanBateman  I agree with everything you say, but there is one advantage to using JVMTI_VERSION_1 instead of JVMTI_VERSION. With the former if you tried, for example, to drop the JDK 21 debug agent into a JDK 20 build, you will get:

`ERROR: This jdwp native library will not work with this VM's version of JVMTI (20.0.0). It needs JVMTI 21.0[.0].`

However, if you use JVMTI_VERSION, the error you will see is:

`ERROR: JDWP unable to access JVMTI Version 21.0.0 (0x30150000). JNIEnv's GetEnv() returned -3.`

This is because the first example passes the GetEnv() but then fails the version check, and the 2nd one fails the GetEnv(). Both are doing the right thing. It's just a matter of the preferable error message, and the 2nd seems better.

**Comment by AlanBateman** [comment]
> @AlanBateman I agree with everything you say, but there is one advantage to using JVMTI_VERSION_1 instead of JVMTI_VERSION. With the former if you tried, for example, to drop the JDK 21 debug agent into a JDK 20 build, you will get:
> 
> `ERROR: This jdwp native library will not work with this VM's version of JVMTI (20.0.0). It needs JVMTI 21.0[.0].`
> 
> However, if you use JVMTI_VERSION, the error you will see is:
> 
> `ERROR: JDWP unable to access JVMTI Version 21.0.0 (0x30150000). JNIEnv's GetEnv() returned -3.`
> 
> This is because the first example passes the GetEnv() but then fails the version check, and the 2nd one fails the GetEnv(). Both are doing the right thing. It's just a matter of the preferable error message, and the 2nd seems better.

I don't have mind. For now (JDK 19 and 20), the case where it will only likely arise is -agentlib:jdwp --enable-preview on ports that don't have VM continuations. In that case, GetEnv will fail. If virtual threads become permanent in JDK 21 and JVMTI is not updated to work with the alternative implementation of virtual threads then we might have to expand the error message a bit.

**Comment by plummercj** [comment]
> I don't mind. For now (JDK 19 and 20), the case where it will only likely arise is -agentlib:jdwp --enable-preview on ports that don't have VM continuations. In that case, GetEnv will fail. If virtual threads become permanent in JDK 21 and JVMTI is not updated to work with the alternative implementation of virtual threads then we might have to expand the error message a bit.

Seems like that is going to be an issue for any JVMTI agent, and I'm not too sure how we would determine the issue is a lack of JVMTI support on the platform. Is seems JVMTI is currently returning EVERSION when it is not supported, which isn't all the useful in determining the underlying reason.

**Review by sspitsyn — APPROVED** [review]
Looks good.
Thanks,
Serguei

**Comment by plummercj** [comment]
I'm going to stick with my change to use JVMTI_VERSION rather than JVMTI_VERSION_1. It seems that the agent really should be asking for the version it needs, not asking for a lesser version. Although the error message might not be as good if someone does placement of the debug agent into a different JDK build, no one should be doing this in the first place, and the resulting error message is still appropriate.

**Comment by plummercj** [comment]
Thanks for the reviews Alan, Serguei, and Alex!

/integrate

**Pull request closed without merging**